### PR TITLE
Github-issue#1048 : s3-sink with local-file buffer implementation.

### PR DESCRIPTION
**Pull request by deepaksahu562**
### Description

- Implementation of s3-sink local-file buffer functionality.
 
### Issues Resolved
GitHub-issue #1048 
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferTypeOptions.java [hunk 1]
```diff
@@ -5,13 +5,34 @@
...
+import java.util.stream.Collectors;
+
 /**
  * Defines all the buffer types enumerations.
  */
 public enum BufferTypeOptions {
 
-    INMEMORY,
-    LOCALFILE
+    LOCALFILE("local_file", new LocalFileBuffer());
```
Comment (kkondaka) [inline data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferTypeOptions.java:18]: I thought the plan is to support both INMEMORY and LOCALFILE. Isn't it?

Comment (venkataraopasyavula) [reply data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferTypeOptions.java:18]: We are supporting both INMEMORY and LOCALFILE buffer type. The INMEMORY buffer functionality was raised as separate PR: https://github.com/opensearch-project/data-prepper/pull/2623

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(LocalFileBuffer.class);
+    private BufferedOutputStream bufferedOutputStream;
+    private int eventCount;
+    private final StopWatch watch;
+    private File localFile;
+
+    LocalFileBuffer() {
+        try {
+            localFile = new File(String.valueOf(UUID.randomUUID()));
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java:36]: We should use a temp file here. It could be in the system default temp, or we can allow a user to configure this directory.

Java provides two [`createTempFile`](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#createTempFile-java.lang.String-java.lang.String-) functions that can help here.

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A buffer can hold local file data and flushing it to S3.
+ */
+public class LocalFileBuffer implements Buffer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LocalFileBuffer.class);
+    private BufferedOutputStream bufferedOutputStream;
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java:29]: I'd recommend making this just an `OutputStream` so that the code doesn't have to change if you change this stream class.

```
private OutputStream outputStream;
```

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /**
+     * write byte array to output stream.
+     * @param bytes byte array.
+     * @throws IOException while writing to output stream fails.
+     */
+    @Override
+    public void writeEvent(byte[] bytes) throws IOException {
+        bufferedOutputStream.write(bytes);
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java:95]: We need some synchronization for these writes. Or the events may end up out-of-order.

We can add it here (and to the `InMemoryBuffer`). Or perhaps we can synchronize in the `S3SinkService` itself.

File: data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        Assertions.assertNotNull(localFileBufferFactory);
+    }
+
+    @Test
+    void test_buffer_notNull() {
+        LocalFileBufferFactory localFileBufferFactory = new LocalFileBufferFactory();
+        Assertions.assertNotNull(localFileBufferFactory);
+        Buffer buffer = localFileBufferFactory.getBuffer();
+        Assertions.assertNotNull(buffer);
+        assertThat(buffer, instanceOf(Buffer.class));
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java:26]: You should assert that this is an instance of `LocalFileBuffer` as well.

```
assertThat(buffer, instanceOf(LocalFileBuffer.class))
```

File: data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    @Test
+    void test_with_write_event_into_buffer() throws IOException {
+        localFileBuffer = new LocalFileBuffer();
+
+        while (localFileBuffer.getEventCount() < 55) {
+            localFileBuffer.writeEvent(generateByteArray());
+        }
+        assertThat(localFileBuffer.getSize(), greaterThan(1l));
+        assertThat(localFileBuffer.getEventCount(), equalTo(55));
+        assertThat(localFileBuffer.getDuration(), greaterThanOrEqualTo(0L));
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferTest.java:36]: You should be able to assert the exact duration here.

```
assertThat(localFileBuffer.getDuration(), equalTo(55 * 1000L));
```

File: data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        assertDoesNotThrow(() -> {
+            localFileBuffer.flushToS3(s3Client, BUCKET_NAME, "log.txt");
+        });
+    }
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferTest.java:61]: You need to verify that the file was flushed to S3. Since this is not an integration test, you will do this via mocking.

It should look something like the following:

```
ArgumentCaptor<PutObjectRequest> putObjectRequestArgumentCaptor = ArgumentCaptor.forClass(PutObjectRequest.class);

verify(s3Client).putObject(putObjectRequestArgumentCaptor.capture(), any(RequestBody.class));

PutObjectRequest actualRequest = putObjectRequestArgumentCaptor.getValue();
// add assertions for the bucket and key values.
```

Also, this should validate that the file was actually deleted. You will do this with real files (not mocks).

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactory.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.accumulator;
+
+public class LocalFileBufferFactory implements BufferFactory {
+    @Override
+    public Buffer getBuffer() {
+        return new LocalFileBuffer();
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactory.java:11]: You can probably improve your testing by choosing the `File` here and passing it in as a parameter.

```
File = File.createTempDir(...);
return new LocalFileBuffer(file);
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java (hunk 2 above)
Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java]: Addressed. Implemented using Java provided functionality.

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java (hunk 3 above)
Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBuffer.java]: Modified.

File: data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java (hunk 5 above)
Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java:28]: Addressed.

File: data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferTest.java (hunk 6 above)
Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferTest.java]: Addressed.

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactory.java (hunk 8 above)
Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactory.java]: Addressed.

File: data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferTypeOptions.java [hunk 9]
```diff
@@ -15,7 +18,16 @@
  */
 public enum BufferTypeOptions {
 
-    INMEMORY("in_memory", new InMemoryBuffer());
+    INMEMORY("in_memory", new InMemoryBuffer()),
+    LOCALFILE("local_file", new Object() {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferTypeOptions.java:22]: Why did you change to make this anonymous class? I think you can continue to make this work with `new LocalFileBufferFactory`. This helps with unit testing.

Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/BufferTypeOptions.java]: Modified with `new LocalFileBufferFactory`.

File: data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java (hunk 5 above)
Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java:26]: @dlvenable, Earlier, I didn't understand. Now I can understand and modify as requested.

Comment (dlvenable) [reply data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java:26]: It should be as simple as changing this line to:

```
assertThat(buffer, instanceOf(LocalFileBuffer.class))
```

Here it is with a little more context.

```
Buffer buffer = localFileBufferFactory.getBuffer();
Assertions.assertNotNull(buffer);
assertThat(buffer, instanceOf(LocalFileBuffer.class))
```

We already know this is a `Buffer`. That is handled statically by Java. But, we want to know that this is returning the `LocalFileBuffer` and not an `InMemoryBuffer` (or some other buffer sub-class).

Comment (deepaksahu562) [reply data-prepper-plugins/s3-sink/src/test/java/org/opensearch/dataprepper/plugins/sink/accumulator/LocalFileBufferFactoryTest.java]: Thanks for your clarification, Modified as you suggested.

**Review by dlvenable — APPROVED** [review]
